xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -672,14 +1048,268 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
       TF_RETURN_IF_ERROR(replace_with_new_cc(expectation, 1));
       TF_RETURN_IF_ERROR(replace_with_new_cc(instr, 2));
 
+      // Update the Custom Call associated with the input transpose orders.
+      transpose_orders.key() = new_custom_call->unique_id();
+      transpose_orders_.insert(std::move(transpose_orders));
+
       VLOG(1)
           << "Expectation and norm factor fused into layer norm Custom Call.";
     }
     return OkStatus();
   }
 
+  // Matches and rewrites the backward graph of layer norm patterns into Custom
+  // Calls to cuDNN when the associated forward graph has been rewritten into a
+  // cuDNN Custom Call.
+  Status MatchLayerNormBackward(HloInstruction* instr) {
+    HloInstruction *custom_call = nullptr, *x = nullptr, *x0 = nullptr,
+                   *dy = nullptr, *scale = nullptr, *fused_expectation,
+                   *fused_expectation0, *fused_norm_factor, *fused_norm_factor0,
+                   *broadcast, *scalar, *x_center, *x_center0, *dscale, *dbias;
+    std::array<HloInstruction*, 3> dyk, scalek;
+    if (Match(instr,
```

Comment:
As discussed offline, this doesn't seem to match the gradient of layer norm defined [by the Triton documentation](https://triton-lang.org/main/getting-started/tutorials/05-layer-norm.html#sphx-glr-getting-started-tutorials-05-layer-norm-py):

> ![image](https://github.com/openxla/xla/assets/6510203/e94be76d-dfc3-4939-a39b-aa66ca7d8c37)

I verified this formula by hand via the chain rule, using pen and paper (this was quite the exercise and took me hours).

I wrote a quick Python script to run the layer norm gradient using three algorithms: The forumula matched by this Match statement, the formula of the Triton link above, and by computing the gradient numerically. The Triton formula matches the nuemric result, indiciating it's correct, but does not match what is matched in this PR:

```python
import numpy as np

def backwards_layernorm_triton_docs(x, w, grad_y, epsilon):
  N = len(x)
  mean = np.mean(x)
  var = np.var(x)
  stdev = np.sqrt(var + epsilon)
  y = (x - mean) / stdev
  t1 = grad_y * w
  t2 = 1 / N * np.dot(y, grad_y * w) * y
  t3 = 1 / N * np.dot(grad_y, w)
  return 1 / np.sqrt(var + epsilon) * (t1 - t2 - t3)


def backwards_layernorm_phillip(x, w, grad_y, epsilon):
  N = len(x)
  mean = np.mean(x)
  var = np.var(x)
  stdev = np.sqrt(var + epsilon)
  f0 = np.sum((x - mean) * w * grad_y)
  f1 = (x - mean) * (2 / N) * (-0.5 * stdev ** 3) * f0
  f2 = 1/stdev * grad_y * w
  return -1/N * (np.sum(f1) + np.sum(f2)) + f1 + f2


def backwards_layernorm_numeric(x, w, grad_y, epsilon):
  delta = 1e-7
  def layer_norm(x):
    mean = np.mean(x)
    var = np.var(x)
    stdev = np.sqrt(var + epsilon)
    return (x - mean) / stdev * w

  y = layer_norm(x)
  grad_x = np.empty(x.shape)
  for i in range(len(x)):
    x2 = x.copy()
    x2[i] += delta
    y2 = layer_norm(x2)
    diff = np.sum(y2 * grad_y) - np.sum(y * grad_y)
    grad_x[i] = diff / delta
  return grad_x


def main():
  x = np.array([1., 2., 3., 4.])
  w = np.array([2., 3., 4., 5.])
  grad_y = np.array([3., 4., 5., 6.])
  epsilon = np.array(1e-8)

  triton_docs = backwards_layernorm_triton_docs(x, w, grad_y, epsilon)
  phillip = backwards_layernorm_phillip(x, w, grad_y, epsilon)
  numeric = backwards_layernorm_numeric(x, w, grad_y, epsilon)
  print(f'triton_docs={triton_docs}')
  print(f'phillip={phillip}')
  print(f'numeric={numeric}')


if __name__ == '__main__':
    main()
```

I'm not sure why this Match statement does not match the Triton docs formula. If this Match statement were wrong, then the cudnn implementation would also be wrong in the exactly same way, given that the tests pass. Perhaps I transcribed this Match statement incorrectly in the Python code above? Any ideas?

Reply:
It looks like there's a small mistake in your transcription of the implementation.

Try f1 = (x - mean) * (2 / N) * (-0.5 / stdev ** 3) * f0.

Reply:
> It looks like there's a small mistake in your transcription of the implementation.
> 
> Try f1 = (x - mean) * (2 / N) * (-0.5 / stdev ** 3) * f0.

Very good catch! Now your layer norm formula matches the Triton version.